### PR TITLE
MODINVSTOR-1105 - Create new field completeUpdatedDate for Instance

### PR DESCRIPTION
**Pull request by obozhko-folio**
[MODINVSTOR-1105](https://issues.folio.org/browse/MODINVSTOR-1105) - Create new field completeUpdatedDate for Instance

### CODE REVIEW

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+     VOLATILE NOT LEAKPROOF
+ AS $BODY$
+ DECLARE
+    createdDate text;
+ BEGIN
+      SELECT (jsonb ->> 'createdDate')::text INTO createdDate
+      FROM ${myuniversity}_${mymodule}.audit_holdings_record
+      WHERE (jsonb -> 'record' ->> 'id')::uuid = OLD.id;
+
+      UPDATE ${myuniversity}_${mymodule}.instance inst SET completeUpdatedDate =
```
Comment (TarasSpashchenko) [inline src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql:49]: You can rewrite this the following way without querying **audit_holdings_record** table. 
 
UPDATE ${myuniversity}_${mymodule}.instance inst SET completeUpdatedDate = now()
WHERE inst.id = OLD.instanceid;

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    VOLATILE NOT LEAKPROOF
+ AS $BODY$
+ DECLARE
+   createdDate text;
+ BEGIN
+     SELECT (jsonb ->> 'createdDate')::text INTO createdDate
+     FROM ${myuniversity}_${mymodule}.audit_item
+     WHERE (jsonb -> 'record' ->> 'id')::uuid = OLD.id;
+
+     UPDATE ${myuniversity}_${mymodule}.instance inst SET completeUpdatedDate =
```
Comment (TarasSpashchenko) [inline src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql:97]: I believe **now()** can be used here as well

**Review by TarasSpashchenko — CHANGES_REQUESTED** [review]

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql (hunk 1 above)
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql]: Done

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql (hunk 2 above)
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql]: Done

File: src/main/resources/templates/db_scripts/oaipmh/addCompleteUpdatedDate.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE ${myuniversity}_${mymodule}.instance
+  ADD COLUMN IF NOT EXISTS
+    completeUpdatedDate timestamp with time zone;
```
Comment (psmagin) [inline src/main/resources/templates/db_scripts/oaipmh/addCompleteUpdatedDate.sql:3]: please use snake case for column names

Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/addCompleteUpdatedDate.sql:3]: Will it be okay if I use completeupdateddate?

Comment (psmagin) [reply src/main/resources/templates/db_scripts/oaipmh/addCompleteUpdatedDate.sql:3]: why not complete_updated_date? any restrictions/limitations?

Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/addCompleteUpdatedDate.sql:3]: No restrictions, but in this case I need to update code in mod-oai-pmh module. Anyway, no problems, I will use complete_updated_date.

**Review by psmagin — APPROVED** [review]

**Comment by psmagin** [comment]
Please update NEWS file

**Comment by obozhko-folio** [comment]
> Please update NEWS file

Updated.

File: src/main/resources/templates/db_scripts/oaipmh/populateWithUpdatedDate.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+UPDATE ${myuniversity}_${mymodule}.instance
+SET complete_updated_date = (jsonb -> 'metadata' ->> 'updatedDate')::timestamp with time zone;
```
[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/populateWithUpdatedDate.sql:2]: Initially complete_updated_date column is always NULL, so I believe it is redundant.

[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/populateWithUpdatedDate.sql]: It was decided to run this script manually, so it was deleted.

[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/populateWithUpdatedDate.sql]: It was decided to run this script manually, so it was deleted.

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    LANGUAGE 'plpgsql'
+    COST 100
+    VOLATILE NOT LEAKPROOF
+ AS $BODY$
+ BEGIN
+     UPDATE ${myuniversity}_${mymodule}.instance inst SET complete_updated_date = NOW()
+     WHERE inst.id IN (
+         SELECT instanceid
+         FROM ${myuniversity}_${mymodule}.holdings_record hold_rec
+         WHERE hold_rec.id = (OLD.jsonb ->> 'holdingsRecordId')::uuid);
```
[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql]: Done.

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+ CREATE OR REPLACE FUNCTION ${myuniversity}_${mymodule}.completeUpdatedDate_for_insert_update_item()
+     RETURNS trigger
+     LANGUAGE 'plpgsql'
+     COST 100
+     VOLATILE NOT LEAKPROOF
+ AS $BODY$
+ BEGIN
+      UPDATE ${myuniversity}_${mymodule}.instance inst SET complete_updated_date =
+          (NEW.jsonb -> 'metadata' ->> 'updatedDate')::timestamp with time zone
```
[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql]: Done.

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      WHERE inst.id IN (
+          SELECT instanceid FROM ${myuniversity}_${mymodule}.holdings_record hold_rec
+          WHERE hold_rec.id = (NEW.jsonb ->> 'holdingsRecordId')::uuid);
```
[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql]: Done.

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+ CREATE OR REPLACE FUNCTION ${myuniversity}_${mymodule}.completeUpdatedDate_for_insert_update_holding()
+     RETURNS trigger
+     LANGUAGE 'plpgsql'
+     COST 100
+     VOLATILE NOT LEAKPROOF
+ AS $BODY$
+ BEGIN
+      UPDATE ${myuniversity}_${mymodule}.instance inst SET complete_updated_date =
+          (NEW.jsonb -> 'metadata' ->> 'updatedDate')::timestamp with time zone
```
[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql]: Done.

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      WHERE inst.id = (NEW.jsonb ->> 'instanceId')::uuid;
```
[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql]: Done.

File: src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+ ON ${myuniversity}_${mymodule}.holdings_record;
+ DROP TRIGGER IF EXISTS completeUpdatedDate_item_inserted_updated
+ ON ${myuniversity}_${mymodule}.item;
+ DROP TRIGGER IF EXISTS completeUpdatedDate_item_deleted
+ ON ${myuniversity}_${mymodule}.item;
+
+ DROP FUNCTION IF EXISTS ${myuniversity}_${mymodule}.completeUpdatedDate_for_insert_update_holding;
+ DROP FUNCTION IF EXISTS ${myuniversity}_${mymodule}.completeUpdatedDate_for_delete_holding;
+ DROP FUNCTION IF EXISTS ${myuniversity}_${mymodule}.completeUpdatedDate_for_insert_update_item;
+ DROP FUNCTION IF EXISTS ${myuniversity}_${mymodule}.completeUpdatedDate_for_delete_item;
```
[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/createTriggersAndFunctionsForCompleteUpdatedDate.sql]: Removed.

File: src/main/resources/templates/db_scripts/oaipmh/modifySetInstanceMdJsonTrigger.sql [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    NEW.jsonb = jsonb_set(NEW.jsonb, '{metadata,updatedDate}', to_jsonb(to_char(updatedDate, 'YYYY-MM-DD"T"HH24:MI:SS.MS"Z"')));
+    -- since this trigger is being executed after triggers on update/insert/delete items/holdings,
+    -- need to check current value to avoid overwriting with the old updatedDate value
+    IF ((NEW.jsonb -> 'metadata' ->> 'updatedDate')::timestamp with time zone > NEW.complete_updated_date) THEN
+    	NEW.complete_updated_date = (NEW.jsonb -> 'metadata' ->> 'updatedDate')::timestamp with time zone;
+    END IF;
+  end if;
+  RETURN NEW;
+END;
+$BODY$;
```
[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/modifySetInstanceMdJsonTrigger.sql]: I created a new trigger updateCompleteUpdatedDate_instance.

File: src/main/resources/templates/db_scripts/oaipmh/addCompleteUpdatedDate.sql [hunk 12]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE ${myuniversity}_${mymodule}.instance
+  ADD COLUMN IF NOT EXISTS
+    complete_updated_date timestamp with time zone;
```
[the remark this answers is not in the document]
Comment (obozhko-folio) [reply src/main/resources/templates/db_scripts/oaipmh/addCompleteUpdatedDate.sql:3]: Unit tests are created.

**Review by TarasSpashchenko — APPROVED** [review]

**Review by psmagin — APPROVED** [review]

**Pull request merged**